cmd/acraserver/acraserver.go

```diff
@@ -175,7 +187,79 @@ func main() {
 		}()
 	}
 	if *withZone && !*disableHTTPApi {
-		go server.StartCommands()
+		if os.Getenv("_GRACEFUL_RESTART") == "true" {
+			go server.StartCommandsFromFileDescriptor(4)
+		} else {
+			go server.StartCommands(true)
+		}
+
+	}
+
+	log.Infof("PID: %v", os.Getpid())
+	if os.Getenv("_GRACEFUL_RESTART") == "true" {
+		go server.StartFromFileDescriptor(3)
+	} else {
+		go server.Start(true)
 	}
-	server.Start()
+
+	signal.Notify(SignalsChannel, syscall.SIGHUP, syscall.SIGTERM)
+	var WaitTimeoutError = errors.New("timeout")
+	for sig := range SignalsChannel {
+		if sig == syscall.SIGTERM {
+			// Stop accepting new connections
+			server.Stop(server.socketACRA)
+			if *withZone && !*disableHTTPApi {
+				server.Stop(server.socketAPI)
+			}
+			// Wait a maximum of 10 seconds for existing connections to finish
+			err := server.WaitWithTimeout(10 * time.Second)
+			if err == WaitTimeoutError {
+				log.Debugf("Timeout when stopping server, %d active connections will be cut.\n", server.ConnectionsCounter())
+				os.Exit(-127)
```

Comment:
why `-127`? What about to add some error message to output? For the case when app will work without debug output

Reply:
Refactored